### PR TITLE
*: partitions interfaces and ResolvedTable node

### PR DESCRIPTION
**Pull request by mcarmonaa**
- Interfaces from #309 #310 #311 has been added.
- A new node `plan.ResolvedTable`has been created.
- `mem.Table` now implements the interfaces from #309 #310 #311
- Now some rules are excuted only once before and after the rest of the `DefaultRules`
- `CreateIndex` now iters over all the table's partitions.
- The pushdown rules now make use of the interfaces from #310 #311 and the `plan.ResolvedTable`, so old nodes from `plan/pushdown.go` have been removed.
- Some rules has been fixed to works with the new interfaces.
- All the test have needed changes to use the `mem.Table`, `plan.ResolvedTable` and the new interfaces.

Note that the specific implementation for the lookups haven't changed and don't comply with the new `IndexLookup` interface to handle partitions, so the code under `sql/index/pilosa` and `sql/index/pilosalib` doesn't compile.

I commented two test in `engine_test.go` that made use of the pilosa indexes. When the pilosa indexes implement the `IndexLookup` interfaces, they should work.

Everything else seems to work properly. :+1:

### CODE REVIEW

File: sql/plan/resolved_table.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+import (
+	"fmt"
+	"io"
+
+	"gopkg.in/src-d/go-mysql-server.v0/sql"
+)
+
+// ResolvedTable represents a resolved SQL Table.
+type ResolvedTable struct {
```
Comment (kuba--) [inline sql/plan/resolved_table.go]: Does it mean that for `ResolvedTable` we keep also `Node` interface?

File: sql/plan/resolved_table.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+	row, err := i.rows.Next()
+	if err != nil && err == io.EOF {
+		i.partition = nil
+		if err := i.rows.Close(); err != nil {
+			return nil, err
+		}
+
+		i.rows = nil
+		return i.Next()
```
Comment (kuba--) [inline sql/plan/resolved_table.go]: tail recursion is fine, I just thought about `goto` ;)

File: sql/plan/resolved_table.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	table      sql.TableV2
+	partitions sql.PartitionIter
+	partition  sql.Partition
+	rows       sql.RowIter
+}
+
+func (i *tableIter) Next() (sql.Row, error) {
+	if i.partition == nil {
+		partition, err := i.partitions.Next()
+		if err != nil {
```
Comment (kuba--) [inline sql/plan/resolved_table.go]: if it's EOF we don't need to close it?

File: sql/plan/resolved_table.go (hunk 2 above)
Comment (mcarmonaa) [reply sql/plan/resolved_table.go]: Is this a case where is not considered a "bad practice" use `goto`?

File: sql/plan/resolved_table.go (hunk 1 above)
Comment (kuba--) [reply sql/plan/resolved_table.go]: No, I think you're right resolved table needs a Node, but in pure table we have to get rid of node.

File: sql/plan/resolved_table.go (hunk 2 above)
Comment (kuba--) [reply sql/plan/resolved_table.go]: No, do not use `goto`. Tail recursion is fast and we won't have huge amount of partitions (I suppose). Personally I just like `goto` - it's fast ;)

File: sql/plan/resolved_table.go (hunk 1 above)
Comment (mcarmonaa) [reply sql/plan/resolved_table.go]: yes, you can see a dummy implementation of  `sql.TableV2` in the test used by the `ResolvedTable` node.

**Comment by ajnavarro** [comment]
@mcarmonaa what is needed to merge this?

File: sql/analyzer/pushdown.go [hunk 4]
```diff
@@ -9,6 +9,27 @@ import (
 	"gopkg.in/src-d/go-mysql-server.v0/sql/plan"
 )
 
+func eraseProjection(ctx *sql.Context, a *Analyzer, node sql.Node) (sql.Node, error) {
```
Comment (erizocosmico) [inline sql/analyzer/pushdown.go]: why is this on pushdown? it's an optimization rule

File: mem/table.go [hunk 5]
```diff
@@ -2,86 +2,279 @@ package mem
...
 
+var _ sql.Table = (*Table)(nil)
+var _ sql.Inserter = (*Table)(nil)
+var _ sql.FilteredTable = (*Table)(nil)
+var _ sql.ProjectedTable = (*Table)(nil)
+var _ sql.IndexableTable = (*Table)(nil)
+
 // NewTable creates a new Table with the given name and schema.
-func NewTable(name string, schema sql.Schema) *Table {
+func NewTable(name string, schema sql.Schema, numPartitions int) *Table {
```
Comment (erizocosmico) [inline mem/table.go]: since most of the calls are with 0 numPartitions, how about having `NewTable` and `NewTablePartitioned` for when you want partitions?

File: mem/table.go [hunk 6]
```diff
@@ -2,86 +2,279 @@ package mem
...
 import (
 	"bytes"
-	"encoding/binary"
+	"encoding/gob"
 	"fmt"
 	"io"
+	"strconv"
 
-	errors "gopkg.in/src-d/go-errors.v1"
+	errors "gopkg.in/src-d/go-errors.v0"
```
Comment (ajnavarro) [inline mem/table.go]: why did you change to previous go-errors version?

Comment (erizocosmico) [reply mem/table.go]: that was probably go-imports' thing

Comment (mcarmonaa) [reply mem/table.go]: It was indeed

**Comment by erizocosmico** [comment]
Tests are failing btw

**Comment by erizocosmico** [comment]
@mcarmonaa can you rebase against `partitions`? I'm working on the exchange rule based on this branch and I need to get the exchange implementation too.

**Comment by erizocosmico** [comment]
Just rebased the PR, adapted the exchange and fixed an issue with inserter. `Insert` method was inside the `Table` interface, which is not how it should be.

It still fails to compile both index driver packages, since they don't support partitions yet.

**Comment by erizocosmico** [comment]
@kuba-- yeah, it's going against a feature branch, so it's not a problem that it does not compile. It's not trivial to make it work with partitions, since it requires a lot of changes. I'm doing that in another PR that uses this as its base.

File: engine_test.go [hunk 7]
```diff
@@ -770,157 +764,157 @@ func TestStarPanic197(t *testing.T) {
...
-	defer func() {
-		time.Sleep(1 * time.Second)
-		done, err := e.Catalog.DeleteIndex("foo", "myidx", true)
-		require.NoError(err)
-		<-done
-
-		require.NoError(os.RemoveAll(tmpDir))
-	}()
-}
+// func TestIndexes(t *testing.T) {
```
Comment (ajnavarro) [inline engine_test.go]: shall we uncomment this code, even if it is failing? just to do not forget about that.

File: mem/database.go [hunk 8]
```diff
@@ -41,6 +41,5 @@ func (d *Database) Create(name string, schema sql.Schema) error {
 	}
 
 	d.tables[name] = NewTable(name, schema)
-
```
Comment (ajnavarro) [inline mem/database.go]: unnecessary modified file.

File: mem/table.go [hunk 9]
```diff
@@ -2,86 +2,284 @@ package mem
...
 		}
 	}
 
-	t.data = append(t.data, row.Copy())
 	return nil
 }
 
-func (t Table) String() string {
+// String implements the sql.Table inteface.
+func (t *Table) String() string {
```
Comment (ajnavarro) [inline mem/table.go]: This is still needed at this level or can be implemented at ResolvedTable level?

File: sql/plan/resolved_table.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	sql.Table
+	name string
+}
+
+var _ sql.Table = (*ResolvedTable)(nil)
+var _ sql.Node = (*ResolvedTable)(nil)
+var _ sql.Nameable = (*ResolvedTable)(nil)
+
+// NewResolvedTable creates a new instance of ResolvedTable.
+func NewResolvedTable(name string, table sql.Table) *ResolvedTable {
```
Comment (ajnavarro) [inline sql/plan/resolved_table.go]: If table has name, why do we need to set a name on resolvedTable too?

File: mem/table.go (hunk 9 above)
Comment (mcarmonaa) [reply mem/table.go]: it can be implemented at ResolvedTable level, but the table implementation can provide a better information since there's no way to know if there are filters, projections or indexes set in the table from the ResolvedTable

File: sql/plan/resolved_table.go (hunk 10 above)
Comment (mcarmonaa) [reply sql/plan/resolved_table.go]: `mem.Table` has name, the interface `sql.Table` hasn't it.

Comment (erizocosmico) [reply sql/plan/resolved_table.go]: Maybe `sql.Table` should have a name

File: engine_test.go (hunk 7 above)
Comment (erizocosmico) [reply engine_test.go]: Don't worry, I am fixing those tests in my PR. If he uncomments that code, the root package will not compile
